### PR TITLE
Contributing walkthrough improvements

### PR DESCRIPTION
**Pull request by kaaloo**
# 🚀 Enhanced Code Quality & Developer Experience

## Summary
This PR introduces comprehensive code quality tooling and CI/CD improvements to establish a robust development workflow with automated linting, formatting, and testing. It also includes critical fixes for the RUNNER component and resolves circular import issues.

### 🔧 Pre-commit Hooks Integration
- **Ruff-based linting & formatting** with automatic fixes
- **Comprehensive quality checks** including:
  - Trailing whitespace cleanup
  - Merge conflict detection
  - Large file prevention
  - YAML/TOML syntax validation
- **Automatic installation** via `just sync` command

### 🏗️ Enhanced CI/CD Pipeline
- **New `checks` job** with complete quality validation
- **Improved caching** with uv lockfile-based caching
- **Test coverage reporting** with Codecov integration
- **Better error reporting** with GitHub-formatted outputs
- **Timeout protection** (10 minutes max)

### 🔧 RUNNER Improvements & Bug Fixes
- **Graceful shutdown handling** for ZMQ workers and main loop
- **Circular import resolution** by reordering imports in clients module
- **Import standardization** across all Python files with automatic sorting
- **Enhanced error handling** and signal management

### 🚀 New Development Commands
```bash
# Complete setup including pre-commit hooks
just sync

# Run full stack locally (API + Runner + Streamlit)
just run [log_level=INFO]

# Code quality commands
just format           # Format code with ruff
just lint [args]      # Lint code with optional args
just test             # Run pytest suite
just publish          # Build and publish package

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -9,31 +9,31 @@ license = { text = "MIT" }
...
+    "aiofiles~=24.1.0",
+    "pyarrow~=19.0.1",
+    "deepeval~=3.4.9",
+    "ragas~=0.2.14",
+    "rapidfuzz~=3.13.0",
+    "ecologits~=0.6.2",
+    "toml~=0.10.2",
     # Front
-    "streamlit==1.40.1",
-    "plotly==5.22.0"
```
Comment (dtrckd) [inline pyproject.toml:36]: Are you sure the upgrade of deepeval does not include breaking changes ? 

I was testing the upgrade to deepeval to support custom judge model with deepeval, following the initiative of @pie-leroy on that matter.

File: justfile [hunk 2]
```diff
@@ -193,3 +194,72 @@ get-experiment expid:
...
+
+publish:
+  uv build
+  uv publish
+
+format:
+  ruff format --config=pyproject.toml .
+
+lint:
+  ruff check --config=pyproject.toml --fix .
```
Comment (dtrckd) [inline justfile:261]: I am not totally confident about the auto fix of ruff as I have seen sometime that it can introduce errors.
For example, in my experience, ruff was removing some needed imports because it considered that there were not used (in an __init__.py file)

What are your opinions on this @kaaloo ?

**Review by dtrckd — CHANGES_REQUESTED** [review]
Would you mind if we disable autoformatting on those folders : 

-  the `notebooks/`  - to keep freedom and flexibility in formatting as  notebooks are mostly pedagogical, we sometime wants custom formatting. 
- the `evalap/api/alembic/versions/` scripts ? - because this is autogenerated by alembic, and reformating them can make review stiff in case of conflicts.

Tell me what you think about it :)

**Comment by kaaloo** [comment]
Thanks for the review @dtrckd !  All great points, I will work on these issues today to improve the PR and ping you then.

File: evalap/api/crud.py [hunk 3]
```diff
@@ -234,7 +234,7 @@ def get_experiments(
     if set_id:
         query = query.filter(models.Experiment.experiment_set_id == set_id)
     if orphan:
-        query = query.filter(models.Experiment.experiment_set_id == None)
```
[the remark this answers is not in the document]
Comment (kaaloo) [reply evalap/api/crud.py:237]: Nice catch!  I'll see what we can do about that.

**Comment by dtrckd** [comment]
Hi @kaaloo , ok let's give it try like this.

**Review by dtrckd — APPROVED** [review]

File: evalap/runners/main.py [hunk 4]
```diff
@@ -55,17 +86,57 @@ def main(worker_url=ZMQ_WORKER_URL, sender_url=ZMQ_SENDER_URL):
...
+
+        # Cleanup
+        logger.info("Cleaning up ZMQ resources...")
+        try:
+            receiver.close(linger=0)
+            distributor.close(linger=0)
+            context.term()
+        except Exception as e:
+            logger.error(f"Error during cleanup: {e}")
+        logger.info("Runner shutdown complete")
```
Comment (dtrckd) [inline evalap/runners/main.py:139]: Hi @kaaloo ,

I just wanted to be sure. This asynchronous code on top of ZMQ  was not easy to setup and I iterated to have somthing that was stable and working. 

Did you test your change and are confident that it will works with no regression or breaks ?

File: evalap/runners/dispatcher.py [hunk 5]
```diff
@@ -25,9 +24,9 @@ def _fix_answer_num_count(db, db_exp, commit=True):
     counts = (
         db.query(
             func.count(models.Answer.id).label("num_try"),
-            func.count(case(((models.Answer.answer != None) & (models.Answer.error_msg == None), 1))).label(
-                "num_success"
-            ),
+            func.count(
+                case(((models.Answer.answer is not None) & (models.Answer.error_msg is None), 1))
```
Comment (dtrckd) [inline evalap/runners/dispatcher.py:28]: @kaaloo , this is another place where the tranformation rule should not apply as it breaks the SQLAlchemy that transpile the python code to SQL requests.

**Review by dtrckd — CHANGES_REQUESTED** [review]

**Comment by dtrckd** [comment]
I have merge main here to resolve conflict. I usually merge and not rebase by default, but my .gitconfig lacked the setting `git config --global pull.rebase false`  :/
